clarity/contracts/arkadiko-auction-engine-v4-1.clar

```diff
@@ -0,0 +1,440 @@
+(define-private (sell-diko (auction-id uint))
+  (let (
+    (auction (get-auction-by-id auction-id))
+    (all-collateral-sold (is-eq (get collateral-amount auction) (get total-collateral-sold auction)))
+    (debt-left (if (>= (get total-debt-burned auction) (get debt-to-raise auction))
+      u0
+      (- (get debt-to-raise auction) (get total-debt-burned auction))
+    ))
+  )
+    ;; Stop if not all collateral sold, no debt to raise or auction not open
+    (asserts! all-collateral-sold (ok u0))
+    (asserts! (> debt-left u0) (ok u0))
+    (asserts! (get-auction-open auction-id) (ok u0))
+    
+    (let (
+      (pair-details (unwrap-panic (unwrap-panic (contract-call? .arkadiko-swap-v2-1 get-pair-details .arkadiko-token .usda-token))))
+      (diko-price (/ (* (get balance-x pair-details) u1100000) (get balance-y pair-details))) ;; 10% extra 
+      (diko-to-mint (/ (* debt-left diko-price) u1000000))
+    )
+      ;; Mint DIKO
+      (try! (as-contract (contract-call? .arkadiko-dao mint-token .arkadiko-token diko-to-mint (as-contract tx-sender))))
+
+      ;; Swap DIKO to USDA
+      (try! (as-contract (contract-call? .arkadiko-swap-v2-1 swap-x-for-y .arkadiko-token .usda-token diko-to-mint u0)))
+
+      ;; Burn USDA
+      (try! (as-contract (contract-call? .arkadiko-dao burn-token .usda-token debt-left (as-contract tx-sender))))
+
+      ;; Swap leftover USDA to DIKO
+      (let ((leftover-usda (unwrap-panic (contract-call? .usda-token get-balance (as-contract tx-sender)))))
+        (try! (as-contract (contract-call? .arkadiko-swap-v2-1 swap-y-for-x .arkadiko-token .usda-token leftover-usda u0)))
+      )
+
+      ;; Burn leftover DIKO
+      (let ((leftover-diko (unwrap-panic (contract-call? .arkadiko-token get-balance (as-contract tx-sender)))))
+        (try! (as-contract (contract-call? .arkadiko-dao burn-token .arkadiko-token leftover-diko (as-contract tx-sender))))
+      )
+
+      ;; Update auction
+      (map-set auctions { id: auction-id } (merge auction {
+        total-debt-burned: (+ (get total-debt-burned auction) debt-left)
+      }))
+
+      (print { type: "auction", action: "sell-diko", data: diko-to-mint })
+      (ok diko-to-mint)
+    )
+  )
+)
+
+
+;; @desc calculate the amount of USDA to burn 
+;; @param auction-id; the ID of the auction for which the collateral amount should be calculated
+;; @param oracle; the oracle implementation that provides the on-chain price
+;; @param liquidation-pool; the pool that holds USDA to use in liquidations
+(define-public (burn-usda-amount
+  (auction-id uint)
+  (oracle <oracle-trait>)
+  (liquidation-pool <liquidation-pool-trait>)
+)
+  (let (
+    (auction (get-auction-by-id auction-id))
+    (debt-left (- (get debt-to-raise auction) (get total-debt-burned auction)))
+    (collateral-left (- (get collateral-amount auction) (get total-collateral-sold auction)))
+    (collateral-price (unwrap-panic (get-collateral-discounted-price oracle auction-id)))
```

Comment:
NTH: it would be great if we can get rid of the `unwrap-panic` here and just use `unwrap!`

Reply:
Updated.